crates/blockchain/blockchain.rs

```diff
@@ -471,190 +490,196 @@ impl Blockchain {
-
-        let mut hashed_address_cache: FxHashMap<H160, H256> = Default::default();
-
-        for updates in rx {
-            let current_length = queue_length.fetch_sub(1, Ordering::Acquire);
-            *max_queue_length = current_length.max(*max_queue_length);
-            let hashed_updates: Vec<_> = updates
-                .into_iter()
-                .map(|u| {
-                    let hashed_address = hashed_address_cache
-                        .entry(u.address)
-                        .or_insert_with(|| keccak(u.address));
-                    (*hashed_address, u)
-                })
-                .collect();
-            state_trie_hash = Self::process_incoming_update_message(
-                &self.storage,
-                &mut state_trie,
-                hashed_updates,
-                &mut storage_updates_map,
-                parent_header,
-                &mut state_updates_map,
-                &mut code_updates,
-                &mut account_states,
-            )?;
-        }
-        let state_updates = state_updates_map.into_iter().collect();
-        let storage_updates = storage_updates_map
-            .into_iter()
-            .map(|(a, (_, s))| (a, s.into_iter().collect()))
-            .collect();
-        let code_updates = code_updates.into_iter().collect();
-
-        Ok(AccountUpdatesList {
-            state_trie_hash,
-            state_updates,
-            storage_updates,
-            code_updates,
+        prefix: Option<H256>,
+    ) -> Result<Trie, StoreError> {
+        Ok(match prefix {
+            Some(account_hash) => {
+                let trie = self.storage.open_storage_trie(
+                    account_hash,
+                    parent_header.state_root,
+                    *EMPTY_TRIE_HASH,
+                )?;
+                let root = trie
+                    .db()
+                    .get(Nibbles::default())?
+                    .filter(|rlp| !rlp.is_empty())
+                    .map(keccak)
+                    .unwrap_or(*EMPTY_TRIE_HASH);
+                self.storage
+                    .open_storage_trie(account_hash, parent_header.state_root, root)?
+            }
+            None => self.storage.open_state_trie(parent_header.state_root)?,
         })
     }
 
-    /// Processes a batch of account updates, applying them to the state trie and storage tries,
-    /// and returns the new state root.
-    #[allow(clippy::too_many_arguments)]
-    fn process_incoming_update_message(
-        storage: &Store,
-        state_trie: &mut Trie,
-        updates: Vec<(H256, AccountUpdate)>,
-        storage_updates_map: &mut StoreUpdatesMap,
+    /// Collapses a root branch node into an extension or leaf node if it has only one valid child.
+    /// Returns None if there are no valid children.
+    ///
+    /// Precondition: the branch node's reference to their children are of type NodeRef::Node
```

Comment:
I see we still try to fetch the node from the DB if it's `NodeRef::Hash`, does this precondition still apply?